asciigraf/asciigraf.py

```diff
@@ -41,52 +47,107 @@ def graph_from_ascii(network_string):
+            | *-|    -> Point(3, 1), Point(3, 2) are neighbours
+            |___|
+             ___
+            |  /|
+            |-* |    -> Point(3, 1), Point(1, 2) are neighbours
+            |___|
+             ___
+            |  -|
+            |---|    -> Point(1, 2), Point(3, 2) are neighbours
+            |___|
+             ___
+            |  -|
+            |--Node| -> Point(1, 2), Point(3, 2) are neighbours
+            |___|
+        """
+        neighbouring_positions = set()
+
+        # first, consider neighbours of our char (e.g. if our char
+        # is '-' then any node or edge char to the left or right
+        # is neighbouring to the char at `pos`)
+        for offset in EDGE_CHAR_NEIGHBOURS[edge_chars[pos]]:
+            neighbour = pos + offset
+            if neighbour in edge_chars or neighbour in node_char_to_node:
+                neighbouring_positions |= {neighbour}
+
+        # second, consider chars to which this char could be a neighbour
+        # (e.g. if the char below is a |, our char neighbours it)
+        for offset, valid_char in ABUTTING.items():
+            if edge_chars.get(pos + offset, " ") == valid_char:
+                neighbouring_positions |= {pos + offset}
+
+        # every edge char should end up with exactly 2 neighbours, or
+        # we have a line that doesn't make sense. the neighbours could either
+        # be an adjacent edge character or a character in a node label
+        n_nodes = len(neighbouring_positions)
+        if n_nodes != 2:
+            error_map = draw({
+                pos: edge_chars[pos]
+                for pos in chain((pos, ), neighbouring_positions)
+                if pos in edge_chars
+            })
             raise InvalidEdgeError(
-                "Too few nodes:\nNetwork String:\n" +
-                "{}\nAffected Edge:\n{}".format(network_string, draw(edge_map))
+                "Too {} nodes:".format("many" if n_nodes > 2 else "few") +
+                "\nNetwork String:\n{}".format(network_string) +
+                "\nAffected Edge:\n{}".format(error_map)
             )
+
+        return tuple(neighbouring_positions)
+
+    edge_char_to_neighbours = {
+        pos: get_abutting_edge_chars(pos)
+        for pos in edge_chars.keys()
+    }
+
+    def build_edge_from_position(starting_char_position):
+        def follow_edge(starting_position, neighbour):
+            if neighbour in node_char_to_node:
+                return (neighbour,)
+            else:
+                a, b = edge_char_to_neighbours[neighbour]
+                next_neighbour = a if b == starting_position else b
+                return (neighbour, ) + follow_edge(neighbour, next_neighbour)
+
+        neighbour_1, neighbour_2 = sorted(edge_char_to_neighbours[pos])
+        positions = list(chain(
+            reversed(follow_edge(pos, neighbour_1)), (pos, ),
+            follow_edge(pos, neighbour_2)
+        ))
+
+        new_edge = dict(
+            points=positions[1:-1],
```

Comment:
Why the slice?

Reply:
`positions` includes the position of the node character at each end of the line. for example, in the below graph, the slice cuts out the position of `o` in `node` and `t` in `other_node`
```
node
  |
  |
other_node
```
hmm. this makes me wonder if maybe it would be a good idea to keep those points in the edge list.